src/main/java/io/appium/java_client/remote/SupportsLocation.java

```diff
@@ -16,19 +16,54 @@
 
 package io.appium.java_client.remote;
 
+import io.appium.java_client.CommandExecutionHelper;
+import io.appium.java_client.ExecutesMethod;
+import io.appium.java_client.Location;
+import io.appium.java_client.MobileCommand;
 import org.openqa.selenium.WebDriver;
-import org.openqa.selenium.html5.Location;
 import org.openqa.selenium.html5.LocationContext;
 import org.openqa.selenium.remote.html5.RemoteLocationContext;
 
-public interface SupportsLocation extends WebDriver, LocationContext {
-    public RemoteLocationContext getLocationContext();
+import java.util.Map;
 
+public interface SupportsLocation extends WebDriver, ExecutesMethod, LocationContext {
+
+    @Deprecated(forRemoval = true)
+    RemoteLocationContext getLocationContext();
+
+    /**
+     * Gets the physical location.
+     *
+     * @return A {@link Location} containing the location information. Returns null if the location is not available.
+     */
     default Location location() {
-        return getLocationContext().location();
+        Map<String, Number> result = CommandExecutionHelper.execute(this, MobileCommand.GET_LOCATION);
+        if (result == null) {
+            return null;
+        }
+        return new Location(
+                convertToDouble(result.get("latitude")),
+                convertToDouble(result.get("longitude")),
+                convertToDouble(result.get("altitude"))
+        );
+    }
+
+    private Double convertToDouble(Number number) {
+        if (number instanceof Double) {
+            return (Double) number;
+        } else if (number instanceof Long) {
+            return number.doubleValue();
+        } else {
+            throw new IllegalArgumentException("Can't convert to double: " + number);
+        }
     }
 
-    default void setLocation(Location location) {
```

Comment:
why this method has been removed. I assume we need to keep both methods for now - the one that works with selenium's location and the other that works with Appium's location unless the first one is deprecated and removed

Reply:
- the method is not removed, it's kept as is, only imports are changed
- Appium's `Location` inherits Selenium's `Location`, thus this method works for both types
- when Selenium's `Location` is removed, this method will be changed to use Appium's `Location`

Reply:
like mentioned above with this strategy (inherited data object) users would still have to update their code if they use the original class imported from selenium. also, the actual deprecation message there won't tell them what to use instead (nobody reads the documentation). 
That is why I think that having own (non-inherited data object) might be a better idea as we can still have a proper deprecation message for the legacy method and users won't have to change anything in their codebase once they switch.